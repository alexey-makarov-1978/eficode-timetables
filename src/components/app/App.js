import React, { useState, useEffect } from "react";
import "./App.css";
import RootSearch from "../rootSearch/RootSearch";
import Plan from "../plan/Plan";
import { getPlan, getLocations } from "../../api/planApi";
import Spinner from "../spinner/Spinner";
import Box from "@material-ui/core/Box";
import { makeStyles } from "@material-ui/core/styles";
import Alert from "@material-ui/lab/Alert";

const useStyles = makeStyles({
  root: {
    maxWidth: 400,
  },
});

function App() {
  const classes = useStyles();

  const [plan, setPlan] = useState(null);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(false);
  const [searchLocations, setSearchLocations] = useState([]);
  const [searchString, setSearchString] = useState("");

  const [coordinates, setCoordinates] = useState({
    oLat: null,
    oLon: null,
    dLat: null,
    dLon: null,
  });

  // called when Search button is clicked
  // state change executes hook that queries HSL using GraphQL
  function searchPlan(origin, destination) {
    if (
      !origin ||
      !destination ||
      origin.label === "" ||
      destination.label === ""
    ) {
      setError(true);
      return;
    }

    setCoordinates({
      oLat: origin.lat,
      oLon: origin.lon,
      dLat: destination.lat,
      dLon: destination.lon,
    });
  }

  // Updates autocomplete options when user types in textbox
  // for origin or destination location
  // executes hook that retrieves possible autocomplete options
  function handleSearchChange(searchText) {
    setSearchString(searchText);
  }

  // retrieves possible autocomplete options
  useEffect(() => {
    if (searchString.length > 0) {
      getLocations(searchString)
        .then((response) => {
          setSearchLocations(response);
        })
        .catch((error) => {
          console.log(error);
        });
    }
  }, [searchString]);

  // queries HSL using GraphQL and retrieves the plan
  useEffect(() => {
    if (coordinates.oLat) {
      setError(false);
      setLoading(true);

      getPlan(
        coordinates.oLat,
        coordinates.oLon,
        coordinates.dLat,
        coordinates.dLon
      )
        .then((response) => {
          setPlan(response);
          setLoading(false);
          setError(false);
        })
        .catch((error) => {
          setLoading(false);
          setError(true);
        });
    }
  }, [coordinates.oLat, coordinates.oLon, coordinates.dLat, coordinates.dLon]);

  return (
    <Box
      boxShadow={3}
      bgcolor="background.paper"
      m={1}
      p={1}
      className={classes.root}
    >
      <div className="App">
        <RootSearch
          onSearch={searchPlan}
          handleInputChange={handleSearchChange}
          locations={searchLocations}
        />
        <br></br>
        <Spinner loading={loading} />
        {error ? (
          <Alert severity="error">
            Network issue or incorrect origin/destination!
          </Alert>
        ) : (
          <></>
        )}
        {plan && !loading && !error ? <Plan data={plan} /> : <></>}
      </div>{" "}
    </Box>
  );
}

export default App;
