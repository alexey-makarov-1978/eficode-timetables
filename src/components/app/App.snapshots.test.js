import React from "react";
import App from "./App";
import renderer from "react-test-renderer";

it("App matching snapshot for root component", () => {
  const tree = renderer.create(<App />);
  expect(tree).toMatchSnapshot();
});
