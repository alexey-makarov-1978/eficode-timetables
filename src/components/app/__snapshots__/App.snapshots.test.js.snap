// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matching snapshot for root component 1`] = `
<div
  className="MuiBox-root MuiBox-root-2 makeStyles-root-1"
>
  <div
    className="App"
  >
    <div
      className="MuiPaper-root MuiCard-root makeStyles-root-3 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        className="MuiCardHeader-root makeStyles-header-4"
      >
        <div
          className="MuiCardHeader-content"
        >
          <span
            className="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
          >
            Reittiopas &gt;
          </span>
        </div>
        <div
          className="MuiCardHeader-action"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M4 16c0 .88.39 1.67 1 2.22V20c0 .55.45 1 1 1h1c.55 0 1-.45 1-1v-1h8v1c0 .55.45 1 1 1h1c.55 0 1-.45 1-1v-1.78c.61-.55 1-1.34 1-2.22V6c0-3.5-3.58-4-8-4s-8 .5-8 4v10zm3.5 1c-.83 0-1.5-.67-1.5-1.5S6.67 14 7.5 14s1.5.67 1.5 1.5S8.33 17 7.5 17zm9 0c-.83 0-1.5-.67-1.5-1.5s.67-1.5 1.5-1.5 1.5.67 1.5 1.5-.67 1.5-1.5 1.5zm1.5-6H6V6h12v5z"
            />
          </svg>
        </div>
      </div>
      <div
        className="MuiCardContent-root"
      >
        <div
          className="MuiGrid-root MuiGrid-container"
        >
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
          >
            <div
              aria-expanded={false}
              aria-owns={null}
              className="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasPopupIcon"
              onClick={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="combobox"
            >
              <div
                className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-disabled Mui-disabled"
                  data-shrink={false}
                  id="undefined-label"
                >
                  Mista?
                </label>
                <div
                  className="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot Mui-disabled Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
                  onClick={[Function]}
                >
                  <input
                    aria-activedescendant={null}
                    aria-autocomplete="list"
                    aria-controls={null}
                    aria-invalid={false}
                    autoCapitalize="none"
                    autoComplete="off"
                    autoFocus={false}
                    className="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused Mui-disabled Mui-disabled MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                    disabled={true}
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onMouseDown={[Function]}
                    required={false}
                    spellCheck="false"
                    type="text"
                    value=""
                  />
                  <div
                    className="MuiAutocomplete-endAdornment"
                  >
                    <button
                      aria-label="Open"
                      className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator Mui-disabled Mui-disabled"
                      disabled={true}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={-1}
                      title="Open"
                      type="button"
                    >
                      <span
                        className="MuiIconButton-label"
                      >
                        <svg
                          aria-hidden={true}
                          className="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M7 10l5 5 5-5z"
                          />
                        </svg>
                      </span>
                    </button>
                  </div>
                  <fieldset
                    aria-hidden={true}
                    className="PrivateNotchedOutline-root-6 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      className="PrivateNotchedOutline-legendLabelled-8"
                    >
                      <span>
                        Mista?
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          >
            <button
              aria-label="change"
              className="MuiButtonBase-root MuiIconButton-root"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M9 3L5 6.99h3V14h2V6.99h3L9 3zm7 14.01V10h-2v7.01h-3L15 21l4-3.99h-3z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-1"
          />
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
          >
            <div
              aria-expanded={false}
              aria-owns={null}
              className="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
              onClick={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="combobox"
            >
              <div
                className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                  data-shrink={false}
                  id="undefined-label"
                >
                  Mihin?
                </label>
                <div
                  className="MuiInputBase-root MuiOutlinedInput-root MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
                  onClick={[Function]}
                >
                  <input
                    aria-activedescendant={null}
                    aria-autocomplete="list"
                    aria-controls={null}
                    aria-invalid={false}
                    autoCapitalize="none"
                    autoComplete="off"
                    autoFocus={false}
                    className="MuiInputBase-input MuiOutlinedInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
                    disabled={false}
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onMouseDown={[Function]}
                    required={false}
                    spellCheck="false"
                    type="text"
                    value=""
                  />
                  <div
                    className="MuiAutocomplete-endAdornment"
                  >
                    <button
                      aria-label="Clear"
                      className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator MuiAutocomplete-clearIndicatorDirty"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={-1}
                      title="Clear"
                      type="button"
                    >
                      <span
                        className="MuiIconButton-label"
                      >
                        <svg
                          aria-hidden={true}
                          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                          />
                        </svg>
                      </span>
                    </button>
                    <button
                      aria-label="Open"
                      className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={-1}
                      title="Open"
                      type="button"
                    >
                      <span
                        className="MuiIconButton-label"
                      >
                        <svg
                          aria-hidden={true}
                          className="MuiSvgIcon-root"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M7 10l5 5 5-5z"
                          />
                        </svg>
                      </span>
                    </button>
                  </div>
                  <fieldset
                    aria-hidden={true}
                    className="PrivateNotchedOutline-root-6 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      className="PrivateNotchedOutline-legendLabelled-8"
                    >
                      <span>
                        Mihin?
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
          <div
            className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
          />
        </div>
      </div>
      <div
        className="MuiCardActions-root makeStyles-search-5 MuiCardActions-spacing"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          disabled={false}
          id="searchButton"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Hae reitti
            <span
              className="MuiButton-endIcon MuiButton-iconSizeMedium"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20.94 11c-.46-4.17-3.77-7.48-7.94-7.94V1h-2v2.06C6.83 3.52 3.52 6.83 3.06 11H1v2h2.06c.46 4.17 3.77 7.48 7.94 7.94V23h2v-2.06c4.17-.46 7.48-3.77 7.94-7.94H23v-2h-2.06zM12 19c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
                />
              </svg>
            </span>
          </span>
        </button>
        <br />
        <br />
      </div>
    </div>
    <br />
  </div>
   
</div>
`;
